lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php

```diff
@@ -123,41 +130,76 @@ public function convertException($message, DriverException $exception)
+     *
+     * @return AbstractPlatform|MariaDb102Platform|MySQL57Platform|MySqlPlatform
+     * @throws DBALException
      */
     public function createDatabasePlatformForVersion($version)
     {
-        if ( ! preg_match('/^(?P<major>\d+)(?:\.(?P<minor>\d+)(?:\.(?P<patch>\d+))?)?/', $version, $versionParts)) {
-            throw DBALException::invalidPlatformVersionSpecified(
-                $version,
-                '<major_version>.<minor_version>.<patch_version>'
-            );
+        if (false !== stripos($version, 'mariadb')
+            && version_compare($this->getMariaDbMysqlVersionNumber($version), '10.2.7', '>=')) {
+            return new MariaDb102Platform();
         }
 
-        if (false !== stripos($version, 'mariadb')) {
-            return $this->getDatabasePlatform();
+        if (false === stripos($version, 'mariadb')
+            && version_compare($this->getOracleMysqlVersionNumber($version), '5.7.9', '>=')) {
+            return new MySQL57Platform();
         }
 
+        return $this->getDatabasePlatform();
+    }
+
+    /**
+     * Get a normalized 'version number' from the server string
+     * returned by Oracle MySQL servers.
+     *
+     * @param string $versionString Version string returned by the driver, i.e. '5.7.10'
+     * @throws DBALException
+     */
+    private function getOracleMysqlVersionNumber(string $versionString) : string
+    {
+        if (!preg_match(self::MYSQL_ORACLE_VERSION_REGEXP, $versionString, $versionParts)) {
+            throw DBALException::invalidPlatformVersionSpecified(
+                $versionString,
+                '<major_version>.<minor_version>.<patch_version>'
+            );
+        }
         $majorVersion = $versionParts['major'];
-        $minorVersion = isset($versionParts['minor']) ? $versionParts['minor'] : 0;
-        $patchVersion = isset($versionParts['patch']) ? $versionParts['patch'] : null;
+        $minorVersion = $versionParts['minor'] ?? 0;
+        $patchVersion = $versionParts['patch'] ?? null;
 
         if ('5' === $majorVersion && '7' === $minorVersion && null === $patchVersion) {
             $patchVersion = '9';
         }
 
-        $version = $majorVersion . '.' . $minorVersion . '.' . $patchVersion;
+        return $majorVersion . '.' . $minorVersion . '.' . $patchVersion;
+    }
 
-        if (version_compare($version, '5.7.9', '>=')) {
-            return new MySQL57Platform();
+    /**
+     * Detect MariaDB server version, including hack for some mariadb distributions
+     * that starts with the prefix '5.5.5-'
+     *
+     * @param string $versionString Version string as returned by mariadb server, i.e. '5.5.5-Mariadb-10.0.8-xenial'
+     * @throws DBALException
+     */
+    private function getMariaDbMysqlVersionNumber(string $versionString) : string
+    {
+        $version = str_replace('5.5.5-', '', $versionString);
+
+        if (!preg_match(self::MYSQL_MARIADB_VERSION_REGEXP, strtolower($version), $versionParts)) {
```

Comment:
Having the regexp defined as a constant makes it harder to maintain this code. The capture group names are defined in the regexp but are used in the `return` part and duplicated in `throw`.

Reply:
Previous review... I revert.